compiler/rustc_builtin_macros/src/format.rs

```diff
@@ -1287,266 +237,671 @@ pub fn expand_preparsed_format_args(
+                    match arg_kind {
+                        FormatArgKind::Normal => {
+                            used[index] = true;
+                            Ok(index)
+                        }
+                        FormatArgKind::Named(_) => {
+                            used[index] = true;
+                            numeric_refences_to_named_arg.push((index, span, used_as));
+                            Ok(index)
+                        }
+                        FormatArgKind::Captured(_) => {
+                            // Doesn't exist as an explicit argument.
+                            invalid_refs.push((index, span, used_as, kind));
+                            Err(index)
+                        }
+                    }
+                } else {
+                    invalid_refs.push((index, span, used_as, kind));
+                    Err(index)
+                }
+            }
+            Name(name, span) => {
+                let name = Symbol::intern(name);
+                if let Some(i) = args[start_of_named_args..]
+                    .iter()
+                    .position(|arg| arg.1.ident().is_some_and(|id| id.name == name))
+                {
+                    let index = start_of_named_args + i;
+                    if !matches!(args[index].1, FormatArgKind::Captured(_)) {
+                        used[index] = true;
+                    }
+                    Ok(index)
+                } else {
+                    let span = span.unwrap_or(fmt_span);
+                    let ident = Ident::new(name, span);
+                    let arg = if is_literal {
+                        ecx.expr_ident(span, ident)
+                    } else {
+                        // For the moment capturing variables from format strings expanded from macros is
+                        // disabled (see RFC #2795)
+                        ecx.struct_span_err(span, &format!("there is no argument named `{name}`"))
+                            .note(format!("did you intend to capture a variable `{name}` from the surrounding scope?"))
+                            .note("to avoid ambiguity, `format_args!` cannot capture variables when the format string is expanded from a macro")
+                            .emit();
+                        DummyResult::raw_expr(span, true)
+                    };
+                    args.push((arg, FormatArgKind::Captured(ident)));
+                    Ok(args.len() - 1)
+                }
+            }
+        };
+        FormatArgPosition { index, kind, span }
     };
 
-    // This needs to happen *after* the Parser has consumed all pieces to create all the spans
-    let pieces = unverified_pieces
-        .into_iter()
-        .map(|mut piece| {
-            cx.verify_piece(&piece);
-            cx.resolve_name_inplace(&mut piece);
-            piece
-        })
-        .collect::<Vec<_>>();
+    let mut template = Vec::new();
+    let mut unfinished_literal = String::new();
```

Comment:
Would using `with_capacity` be worth it here? 🤔

Reply:
The first `.push()` to an empty `Vec` and the first `.push_str()` to an empty `String` will pick a reasonable capacity for that first allocation.